### ISSUE TITLE
Xhtml entities

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#442]
Hi - long time! Great work.

I just synced up to the latest revision after almost a year and am having a bit of difficulty with Xhtml entities, in particular `&apos;`

According to [wikipedia](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Entities_representing_special_characters_in_XHTML) `&apos;` is the only entity that is handled differently between XHTML and HTML.

What is the right way to parse XHTML that contains the `&apos;` entity as well as other XHTML entities? The `XmlEntityService` fails because it only has the five basic entities, and `HtmlEntityService` fails on `&apos;`. I did not find an `XhtmlEntityService`. 

Thanks!

**Comment by corliss** [comment]
I was able to hack around my issues by modifying `HtmlEntityService`. I changed the line containing `apos` to `AddBoth` instead of `AddSingle`, and it worked. However, I don't understand the difference between `AddSingle` and `AddBoth`. Another failing unit test involving `mdash` was worked around in the same way.

**Comment by FlorianRappl** [comment]
The easiest way would be to implement your own entity service and use the one that is the closest (i.e., `HtmlEntityService` in your case) for almost all entities - except the ones that are different.

At this stage I am not sure if I should include such a service in the core. Already the XHTML formatter is probably misplaced. I leave that one open until I made up my mind. Also I introduce an XHTML tag to find issues related to XHTML for later usage.

**Comment by corliss** [comment]
Thanks - I don't mind writing a custom entity service, but I'd like to do it correctly. Could you please explain what should be done? For example, I don't understand `AddBoth`  vs `AddSingle`. 

Which entities would be affected? I don't want to miss out on any. I've already made the hack for `apos`, `mdash` and `theta` - just based on random testing.

If I want to do exhaustive unit tests for this service, how should I do that?

Having said that, can I ask why you don't want such a service in the core? (Again, I don't mind writing it and sending a PR.) Isn't XHTML common enough for Anglesharp to support?

**Comment by corliss** [comment]
I also notice that some entities are listed multiple times with different cases, e.g. `gt`, `GT`, and `Gt` - are these all required? The wikipedia page lists only gt. (As you can tell, I am not super knowledgeable about entities and quite wary of the bugs that may arise.)

**Comment by FlorianRappl** [comment]
The `IEntityProvider` interface _is_ really simple: You only need one method called `GetSymbol`, which takes a string and returns a string. That's all!

Apart from that the HTML specification defines some symbols to be used exclusively with the trailing semicolon (`AddSingle`, only adds the given entity), while others can also be used without that delimiter (`AddBoth`, adds the given entity and the entity minus the trailing character - I think that is also the literal translation of the source code). But this is some implementation detail you surely _don't_ need to care about.

Also its not a question about common or not - JavaScript is also common and is beyond the scope of the core AngleSharp. So is a lot of other stuff. It's just not in the scope.

Regarding your question: Yes, entities are case-sensitive in general, but some of them are actually case-insensitive (due to historic reasons). So: Yes. It is indeed needed (actually the class is generated from the official specification - which makes it easy to stay current when the specification is updated).

**Comment by corliss** [comment]
Thanks - where is the official specification for these entities? All the ones I've seen list only lowercase `gt` so I must be looking at the wrong ones. The ones I look at also don't mention the optional trailing semicolon.

Regarding scope - since AngleSharp aims to be a complete angle brackets parser, I wouldn't expect any JavaScript support. But since it does both XML and HTML, it seems natural to support XHTML as well. Just saying.

**Comment by FlorianRappl** [comment]
Nope - just HTML5 is fully supported. End of story.

It's in the standard specification: https://www.w3.org/TR/html/syntax.html#named-character-references

**Comment by corliss** [comment]
No problem.

So here's the reason I was (and still am) confused. I am using the `XmlTokenizer` to parse XHTML. I initialize it with a `TextSource` containing the source XHTML, and also specify `HtmlEntityService` as the resolver.

Here's the weird part: The source contains a single instance of `&apos;` (with a semicolon.) However, the unmodified `HtmlEntityService` throws a `XmlParseError.CharacterReferenceInvalidCode`. Modifying the `HtmlEntityService` so that the line for `apos;` calls`AddBoth` instead of `AddSingle` works.

This is exactly the opposite of what I'd expect - since AddBoth just adds an extra entry for the semicolon-less version of the entity. Since the entity in the source does have a semicolon, I'd expect `AddSingle` to work as well.

So I wonder if there is something in the parser state machine that could cause this weirdness?

**Comment by FlorianRappl** [comment]
It's the wrong parser.

If you use XML you play by the rules of XML. XML entities _always_ end with a semicolon - as such the semicolon is _not_ transported the entity resolver (as it actually _should_ be - the semicolon is part of the syntax, right?). HTML is more robust here, but as a side-effect needs to transport the semicolon to distinguish between entities that can end with a semicolon and those that don't.

Maybe this should be changed somewhat, but when I implemented it I could not find an elegant way that covers the HTML case without changing the signature of the provider or the way the tokenizer works. If you know a good way then shoot - I am open for suggestions.

I don't know if ending with semicolons (also in XML) would be good. Perhaps there was a reason to exclude this - I need to have a second look at it to evaluate this.

**Comment by corliss** [comment]
I see - so the Xml and Html parsers actually require different behaviors from their resolvers? That means that resolvers do not really implement the same semantic interface. That is, `IEntityProvider` is not simple after all, and cannot be described independent of how it will be used. That is really bad, IMHO. Not only is it not reusable, but developers like me will run into this issue again and again. So at the very least separating the signatures - `IXmlEntityProvider` and `IHtmlEntityProvider` seems necessary.

However, it does seem straightforward to just add a semicolon instead. Here's the current code in `XmlTokenizer`:

```
            var length = StringBuffer.Length - start;
            var content = StringBuffer.ToString(start, length);

            if (numeric)
            {
                var number = numeric ? content.FromHex() : content.FromDec();

                if (number.IsValidAsCharRef())
                {
                    StringBuffer.Remove(start, length);
                    return number.ConvertFromUtf32();
                }
            }
            else
            {
                var entity = _resolver.GetSymbol(content);

                if (!String.IsNullOrEmpty(entity))
                {
                    StringBuffer.Remove(start, length);
                    return entity;
                }
            }
```

It is easy to just move the string creation into both the if and else clauses. In the else clause include the semicolon before calling `ToString()`. (As an added bonus it should be possible to avoid string creation in the 'if' clause entirely since we don't actually need the string after parsing it as a number - we can just parse the `StringBuilder` instead.)

Does that work?

**Comment by FlorianRappl** [comment]
See my comment above: I cannot tell at the moment (the solution is obvious and I think there was _at least_ one argument against it).

**Comment by FlorianRappl** [comment]
On a side note: Still even though you use AngleSharp and care enough to criticize some stuff you never starred the repository. I think this would be the least amount of "support" that a project you rely on deserves. What is your opinion here?

**Comment by corliss** [comment]
I had no idea that starring was important, all you had to do was ask - I wish it was not in a way that sounded as if I deliberately didn't star. I use starring just as a way to remember a project - and I hardly need to be reminded about AngleSharp! But I've gone ahead and starred the project.

As for criticism - I'd like to believe that I've contributed through my constructive suggestions especially last year. I don't remember you thanking for that, but that's ok.

It's important to not take things personally - if your users don't tell you when something doesn't seem right, how will you find out?

**Comment by FlorianRappl** [comment]
Before we have a misunderstanding here: My note was simply to remind you that a star would be appreciated - nothing more. Stars have been introduced as bookmarks, but quickly became kind of the currency on GitHub. Projects with no stars run out of support quickly (its hard to find out that someone is interested in the project when there is nothing going on).

I think the rest of your comment is then on the wrong track - nothing personally anywhere. Regarding the discussions though: You have your opinion and agenda and I have mine. That's fine for both of us. The thing you need to understand is that this is a project done in my rare spare time - as such I cannot follow all directions / use-cases, only the ones I think are valuable for the project (i.e., most of the users). Not following one or the other can have multiple reasons and will rarely (if ever) be personal. 

That all being said: Thanks 👍 for your contributions so far. Some truly have been inspiring and fruitful. Always appreciated!

**Comment by corliss** [comment]
Thanks! And I have come to rely on the amazing quality you have manifested through AngleSharp, and am eager to see what you do next.

On another subject - I will probably open a separate thread for this - I've learned quite a few things about avoiding object creation. AngleSharp uses pooling extensively of course, but have you measured the number of strings allocated? The current issue reminded me of this - a string is being created for the purpose of being looked up in the entity table. The Roslyn code base (think I mentioned this last year) is chock full of inventions I haven't seen anywhere else to avoid such allocations. We should discuss this more.

**Comment by corliss** [comment]
FYI, today I tried the simple workaround of adding a semicolon before calling the `HtmlEntityService` from `XmlTokenizer`:

```
                var entity = _resolver.GetSymbol(content + ";");
```

That was enough to make all my tests pass. In other words, the `HtmlEntityService` works perfectly well for Xhtml as well as far as I can tell, and as expected since Xhtml and Html support exactly the same entities.

I'm not suggesting using the workaround as-is of course - there's a needless extra temporary string created.

**Comment by FlorianRappl** [comment]
The string allocation only for the lookup is a possible optimization; but I looked at this before and it has not proven to be of any cost (even in very large documents this is not where most, i.e., greater 1-2% of the time is spent). Most of the strings allocated are really fresh strings (e.g., for attributes and their values) and actually need to be created.

A string slice may help here, but has a fairly large size compared to a pointer for duplicated strings. Besides the downside of duplicates we also have other problems like the need to care about the state of the text source, which can be manipulated by the specification.

If you find another location where strings are only created for a single lookup call we should have a benchmark. In general a benchmark is the first thing needed to drive a change. I made a lot of experiments over time and found some (or even most) "obvious" performance improvements to be negligible.

**Comment by FlorianRappl** [comment]
Looked at this one right now - I think my biggest concern last time indeed was that `GetSymbol` should really just be called with the entity name (excluding the trailing semicolon if possible). However, as we found out this may be too confusing with the HTML specification.

In order to cover both I changed the implementation to _always_ include the trailing semicolon (thus it is only excluded if the underlying tokenizer accepts / tolerates such "invalid" constructs).
